File: windows/tests/SSMConnect.Workflow.Tests/Fixtures/FixtureRunner.cs
using System.Text.Json;
using SSMConnect.Domain;

namespace SSMConnect.Workflow.Tests.Fixtures;

/// <summary>The observable outcome of running one fixture, ready to assert against.</summary>
public sealed record FixtureRun(
    IReadOnlyList<string> States,
    IReadOnlyList<RecordedCall> Calls,
    PortRecorder Recorder,
    ConnectionWorkflow Workflow);

/// <summary>
/// Executes one conformance fixture against the real <see cref="ConnectionWorkflow"/>, with every
/// port faked from the fixture's declared outcomes.
/// </summary>
public sealed class FixtureRunner
{
    private readonly Fixture _fixture;
    private readonly List<string> _states = [];
    private readonly HashSet<string> _firedEvents = [];
    private readonly HashSet<int> _dispatchedSteps = [];
    private readonly Lock _gate = new();

    private FakeTunnelProvider _tunnels = null!;
    private ConnectionWorkflow _workflow = null!;

    public FixtureRunner(Fixture fixture) => _fixture = fixture;

    public async Task<FixtureRun> RunAsync()
    {
        ConnectionProfile profile = FixtureProfileLoader.Load(_fixture);
        AppSettings settings = BuildSettings();
        var recorder = new PortRecorder(_fixture.Given);
        _tunnels = new FakeTunnelProvider(recorder);

        var sink = new RecordingSink(this);
        _workflow = new ConnectionWorkflow(
            new FakeAuthProvider(recorder),
            new FakeEc2Provider(recorder),
            new FakeSsmProvider(recorder),
            new FakeSecretsProvider(recorder),
            new FakeIdentityProvider(recorder),
            new FakeAgentClient(recorder),
            _tunnels,
            new FakeReadinessProbe(recorder),
            new FakeDcvLauncher(recorder, _fixture.Given.ViewerInstalled),
            new FakeInstanceIdStore(recorder, _fixture.Given.LastInstanceId),
            new InstantDelay(),
            sink,
            profile,
            settings,
            BuildTimeouts());

        // Steps without an afterState run immediately, in order. Steps with one are dispatched from
        // the state callback, which fires synchronously inside the workflow, so a cancelling action
        // lands before the next port call rather than racing it.
        for (int index = 0; index < _fixture.When.Length; index++)
        {
            Step step = _fixture.When[index];
            if (step.AfterState is null)
            {
                MarkDispatched(index);
                Dispatch(step.Action);
                await SettleAsync().ConfigureAwait(false);
            }
        }

        await SettleAsync().ConfigureAwait(false);

        return new FixtureRun([.. _states], recorder.Calls, recorder, _workflow);
    }

    private AppSettings BuildSettings()
    {
        AppSettings settings = AppSettings.Default;
        if (_fixture.Settings is not SettingsOverride over)
        {
            return settings;
        }

        return settings with
        {
            AutoConnect = over.AutoConnect ?? settings.AutoConnect,
            AutoReconnect = over.AutoReconnect ?? settings.AutoReconnect,
            ClipboardAutoClearSeconds = over.ClipboardAutoClearSeconds ?? settings.ClipboardAutoClearSeconds,
        };
    }

    private ConnectionTimeouts BuildTimeouts()
    {
        ConnectionTimeouts timeouts = ConnectionTimeouts.Default with
        {
            // The agent retry loop exists to ride out a tunnel that is not listening yet. Under
            // instant delays a long budget would spin uselessly, so the fixture harness keeps it
            // short; the retry behavior itself is still exercised.
            EnsureSessionAttempts = 3,
        };

        if (_fixture.Harness is not HarnessOverride over)
        {
            return timeouts;
        }

        return timeouts with
        {
            MaxReconnectAttempts = over.MaxReconnectAttempts ?? timeouts.MaxReconnectAttempts,
            EstablishRetryAttempts = over.EstablishRetryAttempts ?? timeouts.EstablishRetryAttempts,
            ReconnectBackoff = over.ReconnectBackoffSeconds is double seconds
                ? TimeSpan.FromSeconds(seconds)
                : timeouts.ReconnectBackoff,
        };
    }

    private void Dispatch(string action)
    {
        switch (action)
        {
            case "onLaunch": _workflow.OnLaunch(); break;
            case "connect": _workflow.Connect(); break;
            case "disconnect": _workflow.Disconnect(); break;
            case "reconnect": _workflow.Reconnect(); break;
            case "stopWorkstation": _workflow.StopWorkstation(); break;
            case "handleSystemWake": _ = _workflow.HandleSystemWakeAsync(); break;
            case "applicationWillTerminate": _ = _workflow.ShutdownAsync(); break;
            default: throw new InvalidOperationException($"Unknown fixture action '{action}'.");
        }
    }

    private bool MarkDispatched(int index)
    {
        lock (_gate)
        {
            return _dispatchedSteps.Add(index);
        }
    }

    /// <summary>
    /// Fires anything a fixture pinned to this state: injected events first, so a tunnel drop is
    /// visible to the monitor the workflow is about to start, then pending actions.
    /// </summary>
    private void OnStateChanged(ConnectionState state)
    {
        string wire = state.Wire();
        lock (_gate)
        {
            _states.Add(wire);
        }

        for (int index = 0; index < _fixture.Given.Events.Length; index++)
        {
            FixtureEvent injected = _fixture.Given.Events[index];
            if (injected.AfterState != wire)
            {
                continue;
            }

            bool firstTime;
            lock (_gate)
            {
                firstTime = _firedEvents.Add($"event-{index}");
            }

            if (firstTime)
            {
                FireEvent(injected);
            }
        }

        for (int index = 0; index < _fixture.When.Length; index++)
        {
            Step step = _fixture.When[index];
            if (step.AfterState == wire && MarkDispatched(index))
            {
                Dispatch(step.Action);
            }
        }
    }

    private void FireEvent(FixtureEvent injected)
    {
        switch (injected.Kind)
        {
            case "tunnelDrop":
                // The main tunnel, not a transient agent tunnel: match on the forwarded local port.
                FakeTunnelHandle? handle = _tunnels.Handles
                    .LastOrDefault(h => !h.Terminated && h.LocalPort == _workflow.Profile.LocalPort);
                handle?.InjectProcessExit(injected.Reason?.ExitCode ?? 1, injected.Reason?.StandardError ?? string.Empty);
                break;

            case "systemWake":
                _ = _workflow.HandleSystemWakeAsync();
                break;

            case "applicationWillTerminate":
                _ = _workflow.ShutdownAsync();
                break;

            default:
                throw new InvalidOperationException($"Unknown fixture event '{injected.Kind}'.");
        }
    }

    /// <summary>
    /// Waits until no further work appears. Every fake completes instantly, so a few consecutive
    /// quiet rounds mean the run has genuinely settled — including any auto-reconnect a tunnel
    /// monitor started after the first check.
    /// </summary>
    private async Task SettleAsync()
    {
        for (int round = 0; round < 3; round++)
        {
            await _workflow.WaitForQuiescenceAsync().ConfigureAwait(false);
            await Task.Delay(15).ConfigureAwait(false);
        }

        await _workflow.WaitForQuiescenceAsync().ConfigureAwait(false);
    }

    private sealed class RecordingSink(FixtureRunner owner) : IEventSink
    {
        public void StateChanged(ConnectionState state) => owner.OnStateChanged(state);

        public void Log(string category, string message)
        {
        }

        public void Notify(ConnectionNotification notification)
        {
        }

        public void PasswordAvailable(string password)
        {
        }
    }
}

/// <summary>Loads the profile a fixture references, applying any declared field overrides.</summary>
public static class FixtureProfileLoader
{
    public static ConnectionProfile Load(Fixture fixture)
    {
        string reference = fixture.Profile.Ref
            ?? throw new InvalidOperationException($"Fixture '{fixture.Id}' has no profile reference.");

        string path = Path.GetFullPath(Path.Combine(FixturePaths.WorkflowsDirectory, reference));
        using JsonDocument document = JsonDocument.Parse(File.ReadAllText(path));
        JsonElement root = document.RootElement;

        Dictionary<string, JsonElement> fields = root.EnumerateObject()
            .ToDictionary(property => property.Name, property => property.Value.Clone());

        if (fixture.Profile.Overrides is JsonElement overrides && overrides.ValueKind == JsonValueKind.Object)
        {
            foreach (JsonProperty property in overrides.EnumerateObject())
            {
                fields[property.Name] = property.Value.Clone();
            }
        }

        return new ConnectionProfile
        {
            Id = Guid.Parse(Text(fields, "id")),
            Name = Text(fields, "name"),
            SsoStartUrl = Text(fields, "ssoStartUrl"),
            SsoRegion = Text(fields, "ssoRegion"),
            AccountId = Text(fields, "accountId"),
            RoleName = Text(fields, "roleName"),
            ResourceRegion = Text(fields, "resourceRegion"),
            InstanceTagKey = Text(fields, "instanceTagKey"),
            InstanceTagValue = Text(fields, "instanceTagValue"),
            LocalPort = Number(fields, "localPort", 8443),
            RemotePort = Number(fields, "remotePort", 8443),
            SecretId = OptionalText(fields, "secretId"),
            AgentRemotePort = OptionalNumber(fields, "agentRemotePort"),
            ConnectMode = OptionalText(fields, "connectMode") switch
            {
                "multiUser" => ConnectMode.MultiUser,
                "singleUser" => ConnectMode.SingleUser,
                _ => null,
            },
        };
    }

    private static string Text(Dictionary<string, JsonElement> fields, string name) =>
        fields.TryGetValue(name, out JsonElement value) && value.ValueKind == JsonValueKind.String
            ? value.GetString() ?? string.Empty
            : string.Empty;

    private static string? OptionalText(Dictionary<string, JsonElement> fields, string name) =>
        fields.TryGetValue(name, out JsonElement value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static int Number(Dictionary<string, JsonElement> fields, string name, int fallback) =>
        fields.TryGetValue(name, out JsonElement value) && value.TryGetInt32(out int parsed) ? parsed : fallback;

    private static int? OptionalNumber(Dictionary<string, JsonElement> fields, string name) =>
        fields.TryGetValue(name, out JsonElement value) && value.TryGetInt32(out int parsed) ? parsed : null;
}

/// <summary>Locates the contract fixtures copied alongside the test assembly.</summary>
public static class FixturePaths
{
    public static string FixturesDirectory { get; } =
        Path.Combine(AppContext.BaseDirectory, "contracts", "fixtures");

    public static string WorkflowsDirectory { get; } = Path.Combine(FixturesDirectory, "workflows");

    public static IReadOnlyList<string> WorkflowFixtureFiles() =>
        Directory.Exists(WorkflowsDirectory)
            ? [.. Directory.GetFiles(WorkflowsDirectory, "*.json").OrderBy(path => path, StringComparer.Ordinal)]
            : [];
}

File: SSMConnectKit/Tests/SSMConnectKitTests/Conformance/FixtureRunner.swift
import Foundation
@testable import SSMConnectDomain
@testable import SSMConnectWorkflow
@testable import SSMConnectAWS
@testable import SSMConnectMacOS
@testable import SSMConnectUI

/// Locates `contracts/` by walking up from this source file, so the fixtures are read from the one
/// copy in the repository. Copying them into the test bundle would let the two drift.
enum FixturePaths {
    static let contractsDirectory: URL = {
        var directory = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while directory.path != "/" {
            let candidate = directory.appendingPathComponent("contracts")
            if FileManager.default.fileExists(atPath: candidate.appendingPathComponent("validate.py").path) {
                return candidate
            }
            directory = directory.deletingLastPathComponent()
        }
        fatalError("Could not locate the contracts/ directory above \(#filePath).")
    }()

    static var workflowsDirectory: URL { contractsDirectory.appendingPathComponent("fixtures/workflows") }

    static func workflowFixtureFiles() -> [URL] {
        let contents = (try? FileManager.default.contentsOfDirectory(
            at: workflowsDirectory, includingPropertiesForKeys: nil)) ?? []
        return contents.filter { $0.pathExtension == "json" }.sorted { $0.path < $1.path }
    }

    static func loadFixture(at url: URL) throws -> Fixture {
        try JSONDecoder().decode(Fixture.self, from: Data(contentsOf: url))
    }
}

/// The observable outcome of one fixture run, ready to assert against.
struct FixtureRun {
    let states: [String]
    let calls: [RecordedCall]
    let recorder: PortRecorder
    let machine: ConnectionStateMachine
}

/// Executes one conformance fixture against the real `ConnectionStateMachine`, with every port
/// faked from the fixture's declared outcomes.
@MainActor
final class FixtureRunner {
    private let fixture: Fixture
    private var states: [String] = []
    private var firedEvents: Set<Int> = []
    private var dispatchedSteps: Set<Int> = []

    private var recorder: PortRecorder!
    private var tunnels: FixtureTunnelProvider!
    private var sink: FixtureEventSink!
    private var machine: ConnectionStateMachine!

    init(fixture: Fixture) { self.fixture = fixture }

    func run() async throws -> FixtureRun {
        let profile = try Self.loadProfile(fixture)
        let recorder = PortRecorder(given: fixture.given)
        let tunnels = FixtureTunnelProvider(recorder: recorder)
        let sink = FixtureEventSink()
        self.recorder = recorder
        self.tunnels = tunnels
        self.sink = sink

        machine = ConnectionStateMachine(
            authProvider: FixtureAuthProvider(recorder: recorder),
            ec2: FixtureEC2Provider(recorder: recorder),
            ssm: FixtureSSMProvider(recorder: recorder),
            tunnel: tunnels,
            secrets: FixtureSecretsProvider(recorder: recorder),
            identity: FixtureIdentityProvider(recorder: recorder),
            agent: FixtureAgentClient(recorder: recorder),
            dcv: FixtureDCVLauncher(recorder: recorder, viewerInstalled: fixture.given.viewerInstalled),
            readiness: FixtureReadinessProbe(recorder: recorder),
            tunnelListener: FixtureTunnelListenerProbe(recorder: recorder),
            instanceIds: FixtureInstanceIdStore(recorder: recorder, seed: fixture.given.lastInstanceId),
            // Never send a real signal. `processIdentifier` is a synthetic number from the fixture
            // and would otherwise name an unrelated live process on the host.
            terminateProcess: { [weak tunnels] pid in
                tunnels?.handles.first { $0.processIdentifier == pid }?.terminateByProcessKill()
            },
            log: ConnectionLog(),
            events: sink,
            profile: profile,
            settings: buildSettings(),
            timeouts: buildTimeouts(),
            maxReconnectAttempts: fixture.harness?.maxReconnectAttempts ?? 3,
            reconnectBackoff: .seconds(fixture.harness?.reconnectBackoffSeconds ?? 5),
            // Every backoff collapses to nothing: fixtures assert ordering, never wall-clock time.
            reconnectSleep: { _ in }
        )
        // The sink is what feeds the fixture its ordered state sequence, and it fires synchronously
        // inside the flow — which is what lets a step pinned to `afterState` land before the next
        // port call instead of racing it.
        sink.onState = { [weak self] state in self?.onStateChanged(state) }

        // Steps with no `afterState` run in order, each settling before the next. Steps that name a
        // state are dispatched from the state callback, which fires synchronously inside the flow,
        // so a cancelling action lands before the next port call instead of racing it.
        for (index, step) in fixture.when.enumerated() where step.afterState == nil {
            if dispatchedSteps.insert(index).inserted {
                dispatch(step.action)
                await settle()
            }
        }
        await settle()

        return FixtureRun(states: states, calls: recorder.calls, recorder: recorder, machine: machine)
    }

    // MARK: - Wiring

    private func buildSettings() -> AppSettings {
        var settings = AppSettings.default
        guard let over = fixture.settings else { return settings }
        settings.autoConnect = over.autoConnect ?? settings.autoConnect
        settings.autoReconnect = over.autoReconnect ?? settings.autoReconnect
        settings.clipboardAutoClearSeconds = over.clipboardAutoClearSeconds ?? settings.clipboardAutoClearSeconds
        return settings
    }

    private func buildTimeouts() -> ConnectionTimeouts {
        var timeouts = ConnectionTimeouts.default
        // The agent retry loop exists to ride out a tunnel that is not listening yet. With instant
        // delays a 15-attempt budget would spin pointlessly; 3 still exercises the retry itself.
        timeouts.ensureSessionAttempts = 3
        timeouts.ensureSessionRetryInterval = .zero

        guard let over = fixture.harness else { return timeouts }
        timeouts.establishRetryAttempts = over.establishRetryAttempts ?? timeouts.establishRetryAttempts
        timeouts.establishRetryBackoff = .zero
        for (stage, seconds) in over.stageTimeoutSeconds ?? [:] {
            let duration = Duration.seconds(seconds)
            switch stage {
            case "authenticate": timeouts.authenticate = duration
            case "resolve":      timeouts.resolve = duration
            case "start":        timeouts.start = duration
            case "ssm":          timeouts.ssm = duration
            case "tunnel":       timeouts.tunnel = duration
            case "dcvReady":     timeouts.dcvReady = duration
            default:
                fatalError("Fixture '\(fixture.id)' overrides unknown stage timeout '\(stage)'.")
            }
        }
        return timeouts
    }

    /// Builds the profile a fixture references, applying its shallow field overrides.
    private static func loadProfile(_ fixture: Fixture) throws -> ConnectionProfile {
        guard let reference = fixture.profile.ref else {
            throw FixtureError.message("Fixture '\(fixture.id)' has no profile reference.")
        }
        // `$ref` is relative to the fixture file, so it starts with "../". Append then standardize:
        // building the URL with `relativeTo:` would not resolve the parent traversal.
        let url = FixturePaths.workflowsDirectory.appendingPathComponent(reference).standardized
        guard case var .object(fields) = try JSONDecoder().decode(JSONValue.self, from: Data(contentsOf: url)) else {
            throw FixtureError.message("Profile '\(reference)' is not a JSON object.")
        }
        for (key, value) in fixture.profile.overrides ?? [:] { fields[key] = value }

        func text(_ key: String) -> String { fields[key]?.stringValue ?? "" }
        func number(_ key: String, _ fallback: Int) -> Int { fields[key]?.intValue ?? fallback }

        return ConnectionProfile(
            id: UUID(uuidString: text("id")) ?? UUID(),
            name: text("name"),
            ssoStartUrl: text("ssoStartUrl"),
            ssoRegion: text("ssoRegion"),
            accountId: text("accountId"),
            roleName: text("roleName"),
            resourceRegion: text("resourceRegion"),
            instanceTagKey: text("instanceTagKey"),
            instanceTagValue: text("instanceTagValue"),
            secretId: fields["secretId"]?.stringValue,
            localPort: number("localPort", 8443),
            remotePort: number("remotePort", 8443),
            connectAction: .dcvViewer,
            connectMode: fields["connectMode"]?.stringValue.flatMap(ConnectMode.init(rawValue:)),
            agentRemotePort: fields["agentRemotePort"]?.intValue
        )
    }

    // MARK: - Driving

    private func dispatch(_ action: String) {
        switch action {
        case "onLaunch":                machine.onLaunch()
        case "connect":                 machine.connect()
        case "disconnect":              machine.disconnect()
        case "reconnect":               machine.reconnect()
        case "stopWorkstation":         machine.stopWorkstation()
        case "handleSystemWake":        Task { await self.machine.handleSystemWake() }
        case "applicationWillTerminate": machine.terminateTunnelForQuit()
        default: fatalError("Unknown fixture action '\(action)'.")
        }
    }

    /// Fires whatever a fixture pinned to this state: injected events first, so a tunnel drop is
    /// visible to the monitor the flow is about to start, then any pending actions.
    private func onStateChanged(_ state: ConnectionState) {
        let wire = state.rawValue
        states.append(wire)

        for (index, event) in fixture.given.events.enumerated()
        where event.afterState == wire && firedEvents.insert(index).inserted {
            fire(event)
        }

        for (index, step) in fixture.when.enumerated()
        where step.afterState == wire && dispatchedSteps.insert(index).inserted {
            dispatch(step.action)
        }
    }

    private func fire(_ event: FixtureEvent) {
        switch event.kind {
        case "tunnelDrop":
            // The main tunnel, not a transient agent tunnel: match on the forwarded local port.
            // Deferred because the monitor is started just *after* `connected` is emitted, and a
            // drop delivered before that would be observed by nobody.
            let port = machine.profile.localPort
            Task { @MainActor in
                self.tunnels.handles.last { $0.isActive && $0.localPort == port }?
                    .injectProcessExit(code: Int32(event.reason?.exitCode ?? 1), stderr: event.reason?.stderr ?? "")
            }
        case "systemWake":
            Task { await self.machine.handleSystemWake() }
        case "applicationWillTerminate":
            machine.terminateTunnelForQuit()
        default:
            fatalError("Unknown fixture event '\(event.kind)'.")
        }
    }

    /// Waits until nothing further happens. Every fake completes instantly, so once the recorded
    /// state and call counts hold still across several rounds the run has genuinely settled —
    /// including an auto-reconnect a tunnel monitor started after the first check.
    private func settle() async {
        var quietRounds = 0
        var lastSignature = signature()
        for _ in 0..<200 {
            await machine.awaitInFlightTask()
            await Task.yield()
            try? await Task.sleep(for: .milliseconds(2))

            let current = signature()
            if current == lastSignature {
                quietRounds += 1
                if quietRounds >= 5 { break }
            } else {
                quietRounds = 0
                lastSignature = current
            }
        }
        await machine.awaitInFlightTask()
    }

    private func signature() -> String { "\(states.count)/\(recorder.calls.count)/\(machine.state.rawValue)" }
}

enum FixtureError: Error, CustomStringConvertible {
    case message(String)

    var description: String {
        if case let .message(text) = self { return text }
        return "Fixture error."
    }
}
